control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -1162,49 +1164,94 @@ func getTokenMetaFromDescription(description string) (map[string]string, error)
+	var (
+		instances []*api.CatalogService
+		errs      error
+	)
 
-	// The nodelist may have changed between this point and when the event was raised
-	// For example, if a pod is evicted because a node has been deleted, there is no guarantee that that node will show up here
-	// query consul catalog for a list of nodes supporting this service
-	// quite a lot of results as synthetic nodes are never deregistered.
-	var nodes []*api.Node
-	filter := fmt.Sprintf(`Meta[%q] == %q `, "synthetic-node", "true")
-	nodes, _, err := apiClient.Catalog().Nodes(&api.QueryOptions{Filter: filter, Namespace: namespaces.WildcardNamespace})
+	// Get the names of services that have the provided k8sServiceName and k8sServiceNamespace in their metadata.
+	// This is necessary so that we can then list the service instances for each Consul service name, which may
+	// not match the K8s service name.
+	services, err := r.servicesForK8SServiceNameAndNamespace(apiClient, k8sServiceName, k8sServiceNamespace)
 	if err != nil {
+		r.Log.Error(err, "failed to get catalog services", "name", k8sServiceName)
 		return nil, err
 	}
 
-	var errs error
-	for _, node := range nodes {
-		var nodeServices *api.CatalogNodeServiceList
-		nodeServices, err := r.serviceInstancesForK8SServiceNameAndNamespace(apiClient, k8sServiceName, k8sServiceNamespace, node.Node)
+	// Query consul catalog for a list of service instances matching the given service names.
+	filter := fmt.Sprintf(`NodeMeta[%q] == %q and ServiceMeta[%q] == %q and ServiceMeta[%q] == %q and ServiceMeta[%q] == %q`,
+		metaKeySyntheticNode, "true",
+		metaKeyKubeServiceName, k8sServiceName,
+		constants.MetaKeyKubeNS, k8sServiceNamespace,
+		metaKeyManagedBy, constants.ManagedByValue)
+	for _, service := range services {
+		var is []*api.CatalogService
+		// Always query the default NS. This ensures that we include mesh gateways, which are always registered to the default NS.
+		// It also ensures that during migrations that enable namespaces, we deregister old service instances in the default NS.
+		// An alternative approach to this dual query would be a service instance list using the wildcard NS, which would also
+		// include instances in Consul namespaces that are no longer in use (e.g. configuration change); this capability does
+		// not currently exist in Consul's catalog API (as of 1.17) and would need to first be added.
+		is, _, err = apiClient.Catalog().Service(service, "", &api.QueryOptions{Filter: filter})
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		} else {
-			serviceList = append(serviceList, nodeServices)
+			instances = append(instances, is...)
+		}
+		// If namespaces are enabled a non-default NS is targeted, also query by target Consul NS.
+		if r.EnableConsulNamespaces {
+			nonDefaultNamespace := namespaces.NonDefaultConsulNamespace(r.consulNamespace(k8sServiceNamespace))
+			if nonDefaultNamespace != "" {
```

Comment:
So for clarification, we're going from N round-trips to fetch services from each node to a max of 6 round-trips. But the tradeoff is that we must do a table scan on the services a max of 6 times (a 6x increase in bexpr cost)?

I'd imagine that this is more efficient in most cases, but have we done any benchmarks / testing to verify it's a significant improvement or that it doesn't put too much CPU load on the server?

Reply:
> So for clarification, we're going from N round-trips to fetch services from each node to a max of 6 round-trips. But the tradeoff is that we must do a table scan on the services a max of 6 times (a 6x increase in bexpr cost)?

Brain dumping for visibility and my own later reference; sorry in advance for the long reply.

In practice, I _believe_ we'll actually be doing less bexp filtering and fewer table scans, because before we'd scan the node table, then visit every segment of the node index on services (1 node and 1 service table scan), and now we're traversing the service table once + matched instances (<= 2 service table scans) w/ similar bexp filters. The likelihood this is cheaper goes up in larger clusters (by node count) and those using multiple non-default namespaces.

#### Today
1) Fetch all nodes: table scan nodes, bexp post-filter on node meta, return all nodes
2) For each result of `1` (N nodes), fetch node services by wildcard NS:
  a) Get node (indexed node ID)
  b) `catalogServiceListByNode`, services indexed by node ID
  c) bexp post-filter on service meta, return node service list

Notably, the service scan here accounts for _all_ namespaces, not just default and possibly one more, as is the case in the new approach.

#### Tomorrow
1) Fetch service names: table scan services by NS (partial index scan), bexp post-filter on node and service meta, return just string names
2) For each result of `1`, fetch service instances by name and NS:
  a) Get service instances, indexed by service name and NS (partial index scan)
  b) `parseServiceNodes` decorate each result, indexed node ID
  c) bexp post-filter results on node and service meta, return service node list

We'll do `1` twice in the worst case (default and non-default target NS), and typically `2` twice per result (service and proxy name). `1` results in a NS-bound services table scan, `2` results in a targeted NS-bound index partial scan by name and a single fetch against node index per instance. In atypical deployment schemes where service name is overridden in some cases and not others, we'd repeat `2` that many more times.

I expect in the majority of cases, this will be less costly on the whole just from the decreased ser/de on the first table scans, let alone the potential for less net scanning, but it's true that the query-side difference depends on the specific cluster, the bexp filters are slightly larger for the new queries, and query resource consumption will shift some.

> have we done any benchmarks / testing to verify it's a significant improvement or that it doesn't put too much CPU load on the server?

Per our conversation offline, yep, I agree it's a good safety check before we merge. This is WIP and I'll post back findings here.

Reply:
Testing locally, have not detected any red flags and overall reconcile speed is orders of magnitude faster due to reduced network requests. Hoping to see confirmation from live cluster test as well once completed.